src/Microsoft.ML.Transforms/SvmLight/SvmLightLoader.cs

```diff
@@ -0,0 +1,851 @@
+        /// to a numeric vector of indices. This class is used in case the indices in the input file are numeric. For the case where the input file contains
+        /// feature names, instead of a <see cref="CustomMappingTransformer{TSrc, TDst}"/>, we use a <see cref="ValueToKeyMappingTransformer"/>.
+        /// </summary>
+        private sealed class Indices
+        {
+            public VBuffer<int> FeatureKeys;
+
+            public static void ParseIndicesToOneBased(IntermediateInput input, Indices output)
+            {
+                var editor = VBufferEditor.Create(ref output.FeatureKeys, input.FeatureKeys.Length);
+                var inputValues = input.FeatureKeys.GetValues();
+                for (int i = 0; i < inputValues.Length; i++)
+                {
+                    if (Conversions.Instance.TryParse(in inputValues[i], out int index) && index > 0)
+                        editor.Values[i] = index - 1;
+                    else
+                        editor.Values[i] = -1;
+                }
+                output.FeatureKeys = editor.Commit();
+            }
+
+            public static void ParseIndicesToZeroBased(IntermediateInput input, Indices output)
+            {
+                var editor = VBufferEditor.Create(ref output.FeatureKeys, input.FeatureKeys.Length);
+                var inputValues = input.FeatureKeys.GetValues();
+                for (int i = 0; i < inputValues.Length; i++)
+                {
+                    if (Conversions.Instance.TryParse(in inputValues[i], out int index) && index >= 0)
+                        editor.Values[i] = index;
+                    else
+                        editor.Values[i] = -1;
+                }
+                output.FeatureKeys = editor.Commit();
+            }
+        }
+
+        /// <summary>
+        /// This class and the <see cref="IntermediateOut"/> class are used by the <see cref="CustomMappingTransformer{TSrc, TDst}"/>
+        /// that maps a vector of indices and a vector of values into a single <see cref="VBuffer{T}"/> of values. When the indices
+        /// originate from the <see cref="ValueToKeyMappingTransformer"/> (in case features are specified by name), <see cref="IntermediateOutKeys"/>
+        /// is used, and when they originate from a <see cref="CustomMappingTransformer{TSrc, TDst}"/> that produces an <see cref="Indices"/>,
+        /// <see cref="IntermediateOut"/> is used.
+        /// </summary>
+        private sealed class IntermediateOutKeys
+        {
+            public VBuffer<uint> FeatureKeys;
+            public VBuffer<float> FeatureValues;
+        }
+
+        private sealed class IntermediateOut
+        {
+            public VBuffer<int> FeatureKeys;
+            public VBuffer<float> FeatureValues;
+        }
+
+        private sealed class Output
+        {
```

Comment:
Is there a use case to (say after inference, or during evaluation) to map these indices back to keys or the original strings? In that case, does it make sense to provide a separate function/helper to extract the keys/strings back?

Reply:
In case of using string keys to load the data, the resulting `IDataView` schema will have slot names annotations for the Features column, so the feature names can be extracted this way.

---
In reply to: [332212446](https://github.com/dotnet/machinelearning/pull/4190#discussion_r332212446) [](ancestors = 332212446)

Reply:
(see in `CreateOutputTransformer` inside the `if (keyIndices)` part).

---
In reply to: [343596251](https://github.com/dotnet/machinelearning/pull/4190#discussion_r343596251) [](ancestors = 343596251,332212446)